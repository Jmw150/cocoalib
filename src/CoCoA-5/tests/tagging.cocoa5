-- An old test about tagged values (reactivated 2020-02-13)

a := tagged(7, "dots");

PrintLn "\n\n---Unexisting record---";
PrintLn a;

PrintLn "\n\n---Wrong type---";
PrintTagged := 3;
PrintLn a;

PrintLn "\n\n---No fields ---";
PrintTagged := Record[];
PrintLn a;

PrintLn "\n\n---Field of wrong type--";
PrintTagged.dots := 3;
PrintLn a;

PrintLn "\n\n---Wrong arity---";
PrintTagged.dots := func () endfunc;
PrintLn a;

PrintLn "\n\n---By ref1---";
PrintTagged.dots := func (ref A, B) endfunc;
PrintLn a;

PrintLn "\n\n---By ref2 --";
PrintTagged.dots := func (A, ref B, opt x) endfunc;
PrintLn a;

PrintLn "\n\n---By ref3---";
PrintTagged.dots := func (ref A, ref B) endfunc;
PrintLn a;

PrintLn "\n\n---Not a proc---";
PrintTagged.dots := func (A, B, opt C) return 47; endfunc;
Try
	PrintLn a;
UponError e Do
	PrintLn "should get here...";
EndTry

PrintLn "\n\n---Finally... a valid impl! ;-) ---";
PrintTagged.dots := func (device, value)
		for i:=1 to value do
			print "." on device;
		endfor
	endfunc;
PrintLn a;

PrintLn "\n\n---Other tests---";
PrintLn "[untagged value]  tag=", tag(1);
PrintLn "[with tag \"one\"]  tag=", tag(tagged(1, "one"));
PrintLn "[double tag ?!?]   tag=", tag(untagged(tagged(tagged(1, "one"), "two")));
//PrintLn "tag=", tag(func () endfunc ());


