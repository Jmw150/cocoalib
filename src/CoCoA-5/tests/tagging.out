

---Unexisting record---
--> WARNING: Cannot find "PrintTagged", so I'm implicitly untagging the value
--> WHERE: at line 6 (column 9) of tagging.cocoa5
--> PrintLn a;
-->         ^
7


---Wrong type---
--> WARNING: The variable "PrintTagged" is not a record, so I'm implicitly untagging the value
--> WHERE: at line 10 (column 9) of tagging.cocoa5
--> PrintLn a;
-->         ^
7


---No fields ---
--> WARNING: The variable "PrintTagged" does not contain a field named "dots", so I'm implicitly untagging the value
--> WHERE: at line 14 (column 9) of tagging.cocoa5
--> PrintLn a;
-->         ^
7


---Field of wrong type--
--> WARNING: PrintTagged.dots does not contain a function, so I'm implicitly untagging the value
--> WHERE: at line 18 (column 9) of tagging.cocoa5
--> PrintLn a;
-->         ^
7


---Wrong arity---
--> WARNING: The function PrintTagged.dots cannot receive two arguments, so I'm implicitly untagging the value
--> WHERE: at line 22 (column 9) of tagging.cocoa5
--> PrintLn a;
-->         ^
7


---By ref1---
--> WARNING: The function PrintTagged.dots expects a by-ref argument, so I'm implicitly untagging the value
--> WHERE: at line 26 (column 9) of tagging.cocoa5
--> PrintLn a;
-->         ^
7


---By ref2 --
--> WARNING: The function PrintTagged.dots expects a by-ref argument, so I'm implicitly untagging the value
--> WHERE: at line 30 (column 9) of tagging.cocoa5
--> PrintLn a;
-->         ^
7


---By ref3---
--> WARNING: The function PrintTagged.dots expects a by-ref argument, so I'm implicitly untagging the value
--> WHERE: at line 34 (column 9) of tagging.cocoa5
--> PrintLn a;
-->         ^
7


---Not a proc---
should get here...


---Finally... a valid impl! ;-) ---
.......


---Other tests---
[untagged value]  tag=
[with tag "one"]  tag=one
[double tag ?!?]   tag=
Got 0 error(s) and 8 warning(s).
